src/github.com/matrix-org/dendrite/mediaapi/writers/download.go

```diff
@@ -0,0 +1,525 @@
+	"strconv"
+	"strings"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/mediaapi/config"
+	"github.com/matrix-org/dendrite/mediaapi/storage"
+	"github.com/matrix-org/dendrite/mediaapi/types"
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/matrix-org/util"
+)
+
+// downloadRequest metadata included in or derivable from an download request
+// https://matrix.org/docs/spec/client_server/r0.2.0.html#post-matrix-media-r0-download
+type downloadRequest struct {
+	MediaMetadata *types.MediaMetadata
+	Logger        *log.Entry
+}
+
+// Validate validates the downloadRequest fields
+func (r *downloadRequest) Validate() *util.JSONResponse {
+	// FIXME: the following errors aren't bad JSON, rather just a bad request path
+	// maybe give the URL pattern in the routing, these are not even possible as the handler would not be hit...?
+	if r.MediaMetadata.MediaID == "" {
+		return &util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound("mediaId must be a non-empty string"),
+		}
+	}
+	if r.MediaMetadata.Origin == "" {
+		return &util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound("serverName must be a non-empty string"),
+		}
+	}
+	return nil
+}
+
+func (r *downloadRequest) jsonErrorResponse(w http.ResponseWriter, res util.JSONResponse) {
+	// Marshal JSON response into raw bytes to send as the HTTP body
+	resBytes, err := json.Marshal(res.JSON)
+	if err != nil {
+		r.Logger.WithError(err).Error("Failed to marshal JSONResponse")
+		// this should never fail to be marshalled so drop err to the floor
+		res = util.MessageResponse(500, "Internal Server Error")
+		resBytes, _ = json.Marshal(res.JSON)
+	}
+
+	// Set status code and write the body
+	w.WriteHeader(res.Code)
+	r.Logger.WithField("code", res.Code).Infof("Responding (%d bytes)", len(resBytes))
+	w.Write(resBytes)
+}
+
+var nTries = 5
+
+// Download implements /download
+// Files from this server (i.e. origin == cfg.ServerName) are served directly
+// Files from remote servers (i.e. origin != cfg.ServerName) are cached locally.
+// If they are present in the cache, they are served directly.
+// If they are not present in the cache, they are obtained from the remote server and
+// simultaneously served back to the client and written into the cache.
+func Download(w http.ResponseWriter, req *http.Request, origin gomatrixserverlib.ServerName, mediaID types.MediaID, cfg *config.MediaAPI, db *storage.Database, activeRemoteRequests *types.ActiveRemoteRequests) {
+	r := &downloadRequest{
+		MediaMetadata: &types.MediaMetadata{
+			MediaID: mediaID,
+			Origin:  origin,
+		},
+		Logger: util.GetLogger(req.Context()),
+	}
+
+	// request validation
+	if req.Method != "GET" {
+		r.jsonErrorResponse(w, util.JSONResponse{
+			Code: 405,
+			JSON: jsonerror.Unknown("request method must be GET"),
+		})
+		return
+	}
+
+	if resErr := r.Validate(); resErr != nil {
+		r.jsonErrorResponse(w, *resErr)
+		return
```

Comment:
You're making life harder for yourself than you need to. It's absolutely critical that we *always* return exactly 1 response to the client. This function makes liberal use of "send response then return", and it's easier than you may think to forget one or the other (leading to sending no response or worse: sending a mix of 2 depending on how your conditionals work out because you forgot a `return`).

We should make it a lot harder for us to fall into this trap, factor it out into a function call e.g:
```go
func Download(things, stuff) {
    r := downloadRequest{things, stuff}
    reader, resErr := doDownload(r)
    if resErr != nil {
        r.jsonErrorResponse(w, *resErr)
        return
    }
    if reader == nil { // should never happen
        r.jsonErrorResponse(w, makeAJSONErrorResponse())
        return
    }
    // send actual bytes to client using the reader
}
```

You may find a `reader` alone is insufficient (maybe you need headers and want to keep `doDownload` free from side-effecting on the response), so you can easily factor this all out into a `downloadResponse` struct which has ` reader` and relevant headers. This makes this code a lot easier to reason about, because you can clearly see each return arg that `doDownload` returns being handled, and the compiler will throw a wobbly if we don't return anything from `doDownload`, so I can safely say that we are always returning something.